Include/internal/pycore_object.h

```diff
@@ -272,17 +272,93 @@ _Py_DECREF_NO_DEALLOC(PyObject *op)
 }
 
 #else
-// TODO: implement Py_DECREF specializations for Py_GIL_DISABLED build
+
 static inline void
 _Py_DECREF_SPECIALIZED(PyObject *op, const destructor destruct)
 {
-    Py_DECREF(op);
+    uint32_t local = _Py_atomic_load_uint32_relaxed(&op->ob_ref_local);
+    if (local == _Py_IMMORTAL_REFCNT_LOCAL) {
+        _Py_DECREF_IMMORTAL_STAT_INC();
+        return;
+    }
+    _Py_DECREF_STAT_INC();
+#ifdef Py_REF_DEBUG
+    _Py_DECREF_DecRefTotal();
+#endif
+    if (_Py_IsOwnedByCurrentThread(op)) {
+        local--;
+        _Py_atomic_store_uint32_relaxed(&op->ob_ref_local, local);
+        if (local == 0) {
+            _Py_MergeZeroLocalRefcount(op);
+        }
+    }
+    else {
+        Py_ssize_t shared = _Py_atomic_load_ssize_relaxed(&op->ob_ref_shared);
+        for (;;) {
+            Py_ssize_t new_shared = shared - (1 << _Py_REF_SHARED_SHIFT);
+            if (_Py_atomic_compare_exchange_ssize(
+                    &op->ob_ref_shared,
+                    &shared,
+                    new_shared)) {
+                if (new_shared == 0) {
```

Comment:
You can only safely deallocate if the reference count is zero and it's merged (i.e., new_shared is `_Py_REF_MERGED`)